### ISSUE TITLE
Decorators confuse pyright

### ISSUE THREAD

**Issue by SobhanMP** [microsoft/pyright#10694]
I couldn't find anything to this effect in existing sources, but I suspect I'm not looking in the right place, sorry.

It seems like decorators confuse pyright but not mypy:

```python
import typing as tp


class Wrapped[**P, T]:
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> T:
        raise NotImplementedError

def inner[**P, T](f: tp.Callable[P, T]) -> Wrapped[P, T]:
    assert False

class K:
    @inner
    def f(self, x: int):
        return x

    def g(self, x: int):
        return x


a = K()
a.g(2)
a.f(2) #fails with x missing
```

**Comment by bschnurr** [comment]
checkout https://microsoft.github.io/pyright/#/typed-libraries?id=annotating-decorators  

not sure maybe this
```python
import typing as tp
from typing import TypeVar, Callable, Any

class Wrapped[**P, T]:
    def __init__(self, func: tp.Callable[P, T]):
        self.func = func
    
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> T:
        raise NotImplementedError


def inner[**P, T](f: tp.Callable[P, T]) -> tp.Callable[P, T]:
    """
    Decorator that wraps the function
    """
    return Wrapped(f)

class K:
    @inner
    def f(self, x: int):
        return x

    def g(self, x: int):
        return x


a = K()
a.g(2)
a.f(2)  # Now works correctly
K.f(a, 2)  # Also works
```



```python
import typing as tp
from typing import TypeVar, Callable, Any, cast

_F = TypeVar("_F", bound=Callable[..., Any])

def inner(f: _F) -> _F:
    """
    Decorator that wraps the function but returns the same type
    """

    # Cast to maintain type compatibility
    ...

class K:
    @inner
    def f(self, x: int):
        return x

    def g(self, x: int):
        return x


a = K()
a.g(2)
a.f(2)  # Now works correctly
```

**Comment by erictraut** [comment]
This is related to an under-specified part of the Python typing spec. There is an active discussion about this topic in the [Python typing forum](https://discuss.python.org/t/when-should-we-assume-callable-types-are-method-descriptors/92938/20), and I'm hopeful that those discussions will converge on a consensus for how type checkers should handle this case. 

You mentioned that it works with mypy. That's not what I'm seeing. Here's [a link](https://mypy-play.net/?mypy=latest&python=3.12&gist=d2988dd6acfa1dd499b2a64918a7a26f) to the mypy playground. You can see that it displays the same error as pyright on the same line.

I'm going to close this issue because I don't have plans to change pyright's behavior unless/until the spec clarifies a behavior here.

**Comment by SobhanMP** [comment]
@bschnurr Thanks for the suggestion, but that does not seem to be applicable for my actual use case.

@erictraut 
> You mentioned that it works with mypy. That's not what I'm seeing. Here's [a link](https://mypy-play.net/?mypy=latest&python=3.12&gist=d2988dd6acfa1dd499b2a64918a7a26f) to the mypy playground. You can see that it displays the same error as Pyright on the same line.

I think I have misconfigured mypy; it is not showing any errors in VS Code.


> I'm going to close this issue because I don't have plans to change pyright's behavior unless/until the spec clarifies a behavior here.

Is it really "correct" that `a.f(2)` and `K.f(a, 2)` are treated differently?

> I'm going to close this issue because I don't have plans to change pyright's behavior unless/until the spec clarifies a behavior here.

Looking forward to a resolution.
